allennlp/semparse/contexts/atis_tables.py

```diff
@@ -0,0 +1,434 @@
+"""
+A ``ConversationContext`` represents the context in which an utterance appears by with the
+valid actions and the grammar
+"""
+
+from collections import defaultdict
+from typing import List, Dict, Set
+import re
+
+from parsimonious.expressions import Sequence, OneOf, Literal
+from parsimonious.grammar import Grammar
+
+TWELVE_TO_TWENTY_FOUR = 1200
+HOUR_TO_TWENTY_FOUR = 100
+HOURS_IN_DAY = 2400
+AROUND_RANGE = 30
+
+"""
+This is the base definition of the SQL grammar written in a simplified sort of
+EBNF notation. The notation here is of the form:
+
+    nonterminal = productions
+
+The first element is the starting symbol. We initialize ``col_ref``, ``table_ref``,
+``table_name``, ``number``, ``string`` to empty productions first, we will fill in
+``col_ref``, ``table_ref``, ``table_name`` based on the dataset and ``number`` and
+``string`` based on the utterances
+"""
+
+SQL_GRAMMAR_STR = r"""
+    statement           = query ws ";" ws
+    query               = (ws "(" ws "SELECT" ws distinct ws select_results ws "FROM" ws table_refs ws where_clause ws ")" ws) /
+                          (ws "SELECT" ws distinct ws select_results ws "FROM" ws table_refs ws where_clause ws)
+                        
+    select_results      = col_refs / agg
+
+    agg                 = agg_func ws "(" ws col_ref ws ")"
+    agg_func            = "MIN" / "min" / "MAX" / "max" / "COUNT" / "count"
+
+    col_refs            = (col_ref ws "," ws col_refs) / (col_ref)
+    table_refs          = (table_name ws "," ws table_refs) / (table_name)
+
+    where_clause        = ("WHERE" ws "(" ws conditions ws ")" ws) / ("WHERE" ws conditions ws)
+    
+    conditions          = (condition ws conj ws conditions) / 
+                          (condition ws conj ws "(" ws conditions ws ")") /
+                          ("(" ws conditions ws ")" ws conj ws conditions) /
+                          ("(" ws conditions ws ")") /
+                          ("not" ws conditions ws ) /
+                          ("NOT" ws conditions ws ) /
+                          condition
+    condition           = in_clause / ternaryexpr / biexpr
+
+    in_clause           = (ws col_ref ws "IN" ws query ws)
+
+    biexpr              = ( col_ref ws binaryop ws value) / (value ws binaryop ws value) / ( col_ref ws "LIKE" ws string)
```

Comment:
I think this grammar can over produce strings outside of valid SQL. eg:

`(value ws binaryop ws value)` could currently be completed as `true + col_ref`, which isn't valid SQL.

This problem applies to many of the `*expr` things, I think possibly as a result of `value` being too broad. This will be kind of tough to fix though, so maybe leave it for a later PR.